Cleanup of asserts and throws under the goto-symex directory

NlightNFotis opened the pull request:
This PR is similar in spirit to #2703 but it targets the `goto-symex/` directory.

Commit:
Replace asserts and throws with INVARIANT etc.

NlightNFotis committed:
Translate asserts and throws to PRECONDITIONS and INVARIANTS

tautschnig commented on src/goto-symex/build_goto_trace.cpp:
```diff
@@ -92,7 +92,9 @@ exprt build_full_lhs_rec(
-    assert(tmp.operands().size()==2);
+    DATA_INVARIANT(
+      tmp.operands().size() == 2,
+      "byte_extract_exprt should have two operands.");
```
Let's please use `to_byte_extract_expr`, and use `byte_extract_exprt`'s API. Otherwise we duplicate consistency checks across the code base and risk getting out-of-sync.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -187,8 +187,8 @@ bool goto_symex_statet::constant_propagation_reference(const exprt &expr) const
-    if(expr.operands().size()!=1)
-      throw "member expects one operand";
+    DATA_INVARIANT(
+      expr.operands().size() == 1, "member_exprt takes one operand.");
```
Let's use `to_member_expr`: `return constant_propagation_reference(to_member_expr(expr).compound();`

tautschnig commented on src/goto-symex/postcondition.cpp:
```diff
@@ -76,20 +76,23 @@ bool postconditiont::is_used_address_of(
   {
-    assert(expr.operands().size()==2);
+    DATA_INVARIANT(
+      expr.operands().size() == 2, "index_exprt takes two operands.");
```
Please use `to_index_expr` and use its API below.

tautschnig commented on src/goto-symex/postcondition.cpp:
```diff
@@ -76,20 +76,23 @@ bool postconditiont::is_used_address_of(
   {
-    assert(expr.operands().size()==1);
+    DATA_INVARIANT(
+      expr.operands().size() == 1, "member_exprt takes one operand.");
```
Please use `return is_used_address_of(to_member_expr(expr).compound(), identifier);`

tautschnig commented on src/goto-symex/postcondition.cpp:
```diff
@@ -76,20 +76,23 @@ bool postconditiont::is_used_address_of(
   {
-    assert(expr.operands().size()==1);
+    DATA_INVARIANT(
+      expr.operands().size() == 1, "dereference_exprt takes one operand.");
```
Please use `return is_used(to_dereference_expr(expr).pointer(), identifier);`

tautschnig commented on src/goto-symex/postcondition.cpp:
```diff
@@ -155,7 +158,8 @@ bool postconditiont::is_used(
     // only do index!
-    assert(expr.operands().size()==1);
+    DATA_INVARIANT(
+      expr.operands().size() == 1, "address_of_exprt takes one operand.");
```
See above: use `to_address_of_expr` etc - I'll stop putting the same comments, it applies several times below.

tautschnig commented on src/goto-symex/slice.cpp:
```diff
@@ -112,7 +112,7 @@ void symex_slicet::slice(symex_target_equationt::SSA_stept &SSA_step)
   symex_target_equationt::SSA_stept &SSA_step)
 {
-  assert(SSA_step.ssa_lhs.id()==ID_symbol);
+  PRECONDITION(SSA_step.ssa_lhs.id() == ID_symbol);
```
I believe this should be true by design.

NlightNFotis replied:
Still, isn't it a good idea to leave it there, even if only for documentation purposes?

tautschnig replied:
Not a big deal, but I think if we were to go down that route we'd have, e.g., any use of an `index_exprt` prefixed by `expr.id() == ID_index`.

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -283,7 +285,7 @@ void goto_symext::symex_assign_typecast(
   // these may come from dereferencing on the lhs
 
-  assert(lhs.operands().size()==1);
+  PRECONDITION(lhs.operands().size() == 1);
```
True by virtue of it being a `typecast_exprt`

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -305,9 +307,7 @@ void goto_symext::symex_assign_array(
-
-  if(lhs.operands().size()!=2)
-    throw "index must have two operands";
+  DATA_INVARIANT(lhs.operands().size() == 2, "index_exprt takes two operands");
```
True for all `index_exprt`

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -368,7 +368,8 @@ void goto_symext::symex_assign_struct_member(
   {
-    assert(lhs_struct.operands().size()==1);
+    DATA_INVARIANT(
+      lhs_struct.operands().size() == 1, "typecast_exprt takes one operand.");
```
Use `to_typecast_expr`

tautschnig reviewed (changes requested):
Essentially the same comments as I've put on other PRs related to the `assert` cleanup: let's please use APIs and thereby lift invariants as much as possible.

NlightNFotis commented:
@tautschnig I have implemented most of the changes you suggested. I found them to be very good suggestions. Just two observations: Don't review it yet, I need to go through the whole PR and see where else I could apply the casts with the integrated checks, and also, for some of the superfluous `PRECONDITION`s, aren't they okay, even if only for documentation purposes (iirc they are not built into the release builds, so no performance overhead for a user)

NlightNFotis commented:
As a general update on this, I'm happy with the changes thus far. It's failing CI on some JBMC regression tests, which are expecting some particular output string to be present, alone, in one line. The exception added here makes the particular errors to be multiline, so the tests are failing. I'm going to fix the tests, but first I'm waiting for #2996 to go in, as it contains some exception classes that I think are more appropriate in a couple of cases. Once that goes in, I'm going to rebase, see the behaviour of the tests, and then I'm going to fix the tests that need fixing based on the error messages coming out of the improved exception classes.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -495,12 +493,12 @@ void goto_symex_statet::rename(
   {
+    DATA_INVARIANT(
+      expr.type().id() == ID_pointer,
+      "type of expression required to be pointer");
```
I think this should just be removed, as should the `DATA_INVARIANT` below. We don't even care whether the type of the expression is a pointer, and the subsequent invariant should be replaced by a use of `to_address_of_expr`

tautschnig commented on src/goto-symex/precondition.cpp:
```diff
@@ -78,18 +78,24 @@ void preconditiont::compute_address_of(exprt &dest)
-    assert(dest.operands().size()==2);
+    DATA_INVARIANT(
+      dest.operands().size() == 2,
+      "index expression expected to have two operands");
```
Should be `to_index_expr`

tautschnig commented on src/goto-symex/precondition.cpp:
```diff
@@ -78,18 +78,24 @@ void preconditiont::compute_address_of(exprt &dest)
-    assert(dest.operands().size()==1);
+    DATA_INVARIANT(
+      dest.operands().size() == 1,
+      "member expression expected to have one operand");
```
`to_member_expr`

tautschnig commented on src/goto-symex/precondition.cpp:
```diff
@@ -78,18 +78,24 @@ void preconditiont::compute_address_of(exprt &dest)
-    assert(dest.operands().size()==1);
+    DATA_INVARIANT(
+      dest.operands().size() == 1,
+      "dereference expression expected to have 1 operand");
```
`to_dereference_expr`

tautschnig commented on src/goto-symex/precondition.cpp:
```diff
@@ -104,12 +110,16 @@ void preconditiont::compute_rec(exprt &dest)
-    assert(dest.operands().size()==1);
+    DATA_INVARIANT(
+      dest.operands().size() == 1,
+      "address_of expression expected to have one operand at this point");
```
`to_address_of_expr`

tautschnig commented on src/goto-symex/precondition.cpp:
```diff
@@ -104,12 +110,16 @@ void preconditiont::compute_rec(exprt &dest)
-    assert(dest.operands().size()==1);
+    DATA_INVARIANT(
+      dest.operands().size() == 1,
+      "dereference expression expected to have one operand at this point");
```
`to_dereference_expr`

tautschnig commented on src/goto-symex/slice.cpp:
```diff
@@ -127,7 +127,7 @@ void symex_slicet::slice_assignment(
   symex_target_equationt::SSA_stept &SSA_step)
 {
-  assert(SSA_step.ssa_lhs.id()==ID_symbol);
+  PRECONDITION(SSA_step.ssa_lhs.id() == ID_symbol);
```
See discussion above: I think this is unnecessary.

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -106,12 +110,15 @@ exprt goto_symext::add_to_lhs(
-    assert(p->id()!=ID_symbol);
-    assert(p->operands().size()>=1);
+    INVARIANT(
+      p->id() != ID_symbol, "expected pointed-to expression to be a symbol");
```
_not_ to be a symbol

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -283,8 +293,7 @@ void goto_symext::symex_assign_typecast(
   // these may come from dereferencing on the lhs
-
-  assert(lhs.operands().size()==1);
+  PRECONDITION(lhs.operands().size() == 1);
```
Unnecessary, this is true by virtue of it being a `typecast_exprt`

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -306,26 +315,25 @@ void goto_symext::symex_assign_array(
-    throw "index must have two operands";
+  DATA_INVARIANT(
+    lhs.operands().size() == 2,
+    "index_exprt takes two operands: first must be array; second must be the "
```
Trivially true for `index_exprt`

tautschnig commented on src/goto-symex/symex_atomic_section.cpp:
```diff
@@ -16,10 +16,9 @@ void goto_symext::symex_atomic_begin(statet &state)
-      state.source.pc->source_location.as_string();
+  INVARIANT(
+    state.atomic_section_id == 0,
+    "we don't support nesting of atomic sections");
```
This should be an exception, it's perfectly possible to write input programs that exhibit this.

tautschnig commented on src/goto-symex/symex_atomic_section.cpp:
```diff
@@ -36,8 +35,7 @@ void goto_symext::symex_atomic_end(statet &state)
 
-  if(state.atomic_section_id==0)
-    throw "ATOMIC_END unmatched"; // NOLINT(readability/throw)
+  INVARIANT(state.atomic_section_id != 0, "ATOMIC_END unmatched");
```
Also an exception, as you can easily write a program that does not have matching begin/end pairs.

tautschnig commented on src/goto-symex/symex_builtin_functions.cpp:
```diff
@@ -532,37 +522,33 @@ void goto_symext::symex_macro(
-  {
-    #if 0
-    exprt new_fc("waitfor", fc.type());
+  PRECONDITION(identifier == CPROVER_MACRO_PREFIX "waitfor");
```
Check with @mgudemann @kroening whether we can get rid of the code here (all `#if 0` below).

tautschnig commented on src/goto-symex/symex_dead.cpp:
```diff
@@ -23,11 +23,8 @@ void goto_symext::symex_dead(statet &state)
-
-  if(code.op0().id()!=ID_symbol)
-    throw "dead expects symbol as first operand";
+  PRECONDITION(code.operands().size() == 1);
```
Use `to_code_dead`

tautschnig commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -155,6 +156,9 @@ exprt goto_symext::address_arithmetic(
   {
+    DATA_INVARIANT(
+      expr.operands().size() == 1,
+      "dereference expression expected to have one operand");
```
`to_dereference_expr`

tautschnig commented on src/goto-symex/symex_dereference.cpp:
```diff
@@ -233,8 +238,9 @@ void goto_symext::dereference_rec(
-      throw "dereference takes one operand";
+    DATA_INVARIANT(
+      expr.operands().size() == 1,
+      "dereference expression expected to have one operand");
```
`to_dereference_expr`

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -345,8 +345,9 @@ void goto_symext::merge_goto(
-    throw "atomic sections differ across branches";
+  INVARIANT(
+    state.atomic_section_id == goto_state.atomic_section_id,
+    "atomic sections differ across branches");
```
Should be an exception as one can craft programs exhibiting this.

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -107,8 +108,12 @@ void goto_symext::rewrite_quantifiers(exprt &expr, statet &state)
-    PRECONDITION(expr.op0().id()==ID_symbol);
+    DATA_INVARIANT(
+      expr.operands().size() == 2,
+      "for_all expression expected to have two operands");
```
`to_quantifier_expr`

tautschnig commented on src/goto-symex/symex_start_thread.cpp:
```diff
@@ -18,18 +18,17 @@ void goto_symext::symex_start_thread(statet &state)
+  INVARIANT(
+    state.atomic_section_id == 0,
+    "spawning threads out of atomic sections is not allowed; "
+    "this would require amendments to ordering constraints");
```
Should be an exception as you can write input programs doing so.

NlightNFotis commented:
@tautschnig Hi Michael, I have addressed all the comments in your review, aside from one.

You have suggested to remove some code, after advising with either Matthias or Daniel. It's not easy for me to get hold of either of them now, so I suggest we leave this for a later cleanup (I have already marked it on my todo, so it's not lost, however if you believe that it should be done now, I will see what I can do, I just thought it could be a good idea to not get hung up on this now since this work is relatively high priority).

I want a little bit of advising on a different part of the PR however. In `goto_symex_state.cpp:514` there's a `DATA_INVARIANT` that seems suspiciously general. I thought it a good idea to move it to the transfer functions (`to_if_expr`), but in doing so some of the tests get broken. This happens because all the tests that use our internal malloc get to code that looks like this:

```
__CPROVER_deallocated =
    (malloc_res == __CPROVER_deallocated) ? 0 : __CPROVER_deallocated;
```

where `__CPROVER_deallocated` is declared as a pointer in a different file. (For a live example of this take a look [here](https://github.com/diffblue/cbmc/blob/0c8eff6f3366acf81a7b71a40e893a56aaac2d74/src/ansi-c/library/stdlib.c#L76). In this case `__CPROVER_deallocated`'s type is a `symbol` whereas the `0` is a constant, hence the `DATA_INVARIANT` failing in multiple of the tests. This seems to me that the promotion is not being handled correctly (I have also checked with the C standard `n1124` and that should be typechecking, but on the condition that both the consequent and the alternative expressions are of the same type (in this case `6.5.15-3` indicates that it should typecheck on the condition that `one operand is a pointer and the other is a null pointer constant`).

This made me instinctively want to go and modify the parser, to do automatic implicit typecasts when it parses a conditional expression, however most of what I have tried seems to be convoluted, and most likely wrong (couldn't get them to work). Do you have any particular input on that, or maybe some pointers on what should change? If it's elaborate work, I'm happy to do it as part of a different PR and get this in as it is now if you think it's good as a cleanup PR, I just thought that moving the invariant to the transfer functions is a good piece of work in line with us not duplicating checks.

tautschnig commented:
> [...] In `goto_symex_state.cpp:514` there's a `DATA_INVARIANT` that seems suspiciously general. I thought it a good idea to move it to the transfer functions (`to_if_expr`), but in doing so some of the tests get broken. [...]

This sound like a very good idea, and we should dig deeper. Would you mind creating an fresh PR doing just this (CI will fail, of course). Let's then you and I debug this - it seems that symbol types aren't uniformly expanded.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 8558c578209f9736146411bf525c677b07b5ee86).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85748730
